pkg/querier/worker/scheduler_processor.go

```diff
@@ -163,13 +174,44 @@ func (sp *schedulerProcessor) querierLoop(c schedulerpb.SchedulerForQuerier_Quer
 	}
 }
 
-func (sp *schedulerProcessor) runRequest(ctx context.Context, logger log.Logger, queryID uint64, frontendAddress string, statsEnabled bool, request *httpgrpc.HTTPRequest) {
+func (sp *schedulerProcessor) runQueryRequest(ctx context.Context, logger log.Logger, queryID uint64, frontendAddress string, statsEnabled bool, request *queryrange.QueryRequest) {
+	var stats *querier_stats.Stats
+	if statsEnabled {
+		stats, ctx = querier_stats.ContextWithEmptyStats(ctx)
+	}
+
+	response := handleQueryRequest(ctx, request, sp.handler)
+
+	logger = log.With(logger, "frontend", frontendAddress)
+
+	// Ensure responses that are too big are not retried.
+	if response.Size() >= sp.maxMessageSize {
```

Comment:
This feels like it should be middleware...?

Reply:
Ideally. However, the responses in `queryrangebase.Middleware` have no size method. The inceptors for gRPC have none either :/.